crates/bevy_reflect/bevy_reflect_derive/src/impls/enums.rs

```diff
@@ -262,126 +261,91 @@ struct EnumImpls {
 fn generate_impls(reflect_enum: &ReflectEnum, ref_index: &Ident, ref_name: &Ident) -> EnumImpls {
     let bevy_reflect_path = reflect_enum.meta().bevy_reflect_path();
 
-    let mut variant_info: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_field: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_field_at: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_index_of: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_name_at: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_field_len: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_variant_name: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_variant_type: Vec<proc_macro2::TokenStream> = Vec::new();
+    let mut variant_info = Vec::new();
+    let mut enum_field = Vec::new();
+    let mut enum_field_at = Vec::new();
+    let mut enum_index_of = Vec::new();
+    let mut enum_name_at = Vec::new();
+    let mut enum_field_len = Vec::new();
+    let mut enum_variant_name = Vec::new();
+    let mut enum_variant_type = Vec::new();
 
     for variant in reflect_enum.active_variants() {
         let ident = &variant.data.ident;
         let name = ident.to_string();
         let unit = reflect_enum.get_unit(ident);
 
-        match &variant.fields {
-            EnumVariantFields::Unit => {
-                variant_info.push(quote! {
-                    #bevy_reflect_path::VariantInfo::Unit(
-                        #bevy_reflect_path::UnitVariantInfo::new(#name)
-                    )
-                });
-                enum_variant_name.push(quote! {
-                    #unit => #name
-                });
-                enum_variant_type.push(quote! {
-                    #unit => #bevy_reflect_path::VariantType::Unit
-                });
-            }
-            EnumVariantFields::Unnamed(fields) => {
+        // This is equivalent to a |fields: &Vec<StructField>, to_run: |usize, usize, &StructField| -> TokenStream|
+        // closure. Issue is that the closure cannot itself accept another closure, because closures cannot vary in
+        // the concrete type they accept, and each closure has a different concrete type.
+        macro_rules! for_fields {
+            ($fields:expr, |$reflect_idx:ident, $declar_field:pat, $field:ident| $body:expr) => {{
                 let mut field_info = Vec::new();
-                let mut field_idx: usize = 0;
-                for field in fields.iter() {
-                    if field.attrs.ignore {
+                let mut $reflect_idx: usize = 0;
+                for ($declar_field, $field) in $fields.iter().enumerate() {
```

Comment:
Let's not touch the "reflect vs declaration index" issue here (or elsewhere). I don't want the way indices work here to differ from how they work everywhere else. That deserves it's own PR imo.

I also think that if we did want to handle this properly, we should add it directly to `StructField` rather than having to manage it manually wherever we need it (probably via methods like `reflect_index()` and `decl_index()`).

Reply:
You have
```rust
#[derive(Reflect)]
struct Foo(
  u32, // :0
  #[reflect(ignore)] f32, // :1
  String, // :2
);
```
When ignoring the second field, in the "Reflect world", `Foo` has in practice two fields `u32` and `String`.
Meaning that when you do:
- `<Foo as TupleStruct>::fied(foo, 0)` you get the `u32`
- `<Foo as TupleStruct>::fied(foo, 1)` you get the `String`

For you to get the `String`, you need to keep track of the actual field index, and access it when the "reflect" field index is requested. If you don't do that, when doing `foo.field(1)`, you'd get the `f32`, and when doing `foo.field(2)`, you'd get a `None`. Which seems to be the exact opposite of what the `#[reflect(ignore)]` is supposed to do.

I need to look into it, but if the other tuple struct implementations do not do it that way, it seems pretty buggy.

Reply:
Yeah I think it's a bug. What I'm saying is, I think it would be best if we left the bug in for now and address it in a separate PR (one that also fixes it for all the other traits).

That way, in the interim, we don't have conflicting interpretations of what an index points to. And this also removes the possible debate around it (throw an error vs print a warning vs map the indices vs whatever) from the enum PR.

Reply:
Fair enough. Good reasoning. I'll even open the issue on the bevy repo for you :P. But I still hate the concept of voluntarily introducing bugs :grimacing:

Reply:
So I did a bit of testing. And the current `TupleStruct` trait implementation for `field` does properly account for ignored fields. It does it by using the `StructField.index` field. The `FromReflect` implementation does indeed not work, but we are not implementing `FromReflect` here. The way this PR does thing is the current behavior. If we are to keep it consistent, we should do it the way we do it here, there is no reasons to introduce that bug willingly.

Reply:
Sounds good. We can leave the fixed version in this one and come back to clean it up and fix all the others in a separate PR.